### ISSUE TITLE
About the results of Clip surgery with SAM

### ISSUE THREAD

**Issue by seventy7796** [xmed-lab/CLIP_Surgery#2]
1. I run the demo with CLIP "CS-ViT-B/16", and the results of "person" is as following with wrong points: 
![image](https://user-images.githubusercontent.com/78421476/232498455-e504da46-3e6d-42dd-9620-d5935d180d8d.png)

Could you help me find my errors?

2. More, I try "a person on the bench" and other sentences. Does the results with a long sentence input or with some extra description not work?
3. Could you share the training dataset of "CS-ViT-B/16"? 

Thanks a lot.

**Comment by Eli-YiLi** [comment]
Note that, the demo gives a label set, and CLIP Surgery computes the redundant features from this label set.

I guess you just use a sentence without a label set. In such condition, you should follow the example of for a single text:

### CLIP Surgery for a single text, without fixed label sets
texts = ['shoes']

with torch.no_grad():
    # clip architecture surgery acts on the image encoder
    image_features = model.encode_image(image)
    image_features = image_features / image_features.norm(dim=1, keepdim=True)

    # prompt ensemble for text features with normalization
    text_features = clip.encode_text_with_prompt_ensemble(model, texts, device)

    # extract redundant features from an empty string
    redundant_features = clip.encode_text_with_prompt_ensemble(model, [""], device)

    # apply feature surgery for single text
    similarity = clip.clip_feature_surgery(image_features, text_features, redundant_features)
    similarity_map = clip.get_similarity_map(similarity[:, 1:, :], cv2_img.shape[:2])

This example use the empty string as a redundant feature, and then you can use the similarity map to get points for SAM.

Your attempt also shows the necessity of feature surgery, without it the noises lead to false points.

For multiple words, I will try it.

**Comment by seventy7796** [comment]
@Eli-YiLi 

Thank you for your reply. I follow your code and get the new results which are still not as good as yours.
`all_texts = ['a person in the bench','person', 'bench']
target_texts = ['a person in the bench','person', 'bench']
sam_checkpoint = "sam_vit_h_4b8939.pth"
model_type = "vit_h"
sam = sam_model_registry[model_type](checkpoint=sam_checkpoint)
sam.to(device=device)
predictor = SamPredictor(sam)
predictor.set_image(np.array(pil_img))
with torch.no_grad():
    # clip architecture surgery acts on the image encoder
    image_features = model.encode_image(image)
    image_features = image_features / image_features.norm(dim=1, keepdim=True)

    # prompt ensemble for text features with normalization
    text_features = clip.encode_text_with_prompt_ensemble(model, all_texts, device)

    # extract redundant features from an empty string
    redundant_features = clip.encode_text_with_prompt_ensemble(model, [""], device)

    # apply feature surgery for single text
    similarity = clip.clip_feature_surgery(image_features, text_features, redundant_features)[0]
    # inference SAM with points from CLIP Surgery
    for n in range(similarity.shape[-1]):
        if all_texts[n] not in target_texts:
            continue
        print('similarity',similarity.shape)
        points, labels = clip.similarity_map_to_points(similarity[1:, n], cv2_img.shape[:2], t=0.8)
        masks, scores, logits = predictor.predict(point_labels=labels, point_coords=np.array(points), multimask_output=True)
        mask = masks[np.argmax(scores)]
        mask = mask.astype('uint8')`

text1: person
![image](https://user-images.githubusercontent.com/78421476/232690814-3a1d6f6f-2e21-4280-8728-274484a0ce9c.png)

text2: a person on the bench
![image](https://user-images.githubusercontent.com/78421476/232690881-267ae9a3-ac1e-4603-8ae9-54671b9bed04.png)

text3: bench
![image](https://user-images.githubusercontent.com/78421476/232690936-0763f55a-c36a-4c25-beed-8b923b467457.png)

**Comment by Eli-YiLi** [comment]
I update the demo.ipynb: https://github.com/xmed-lab/CLIP_Surgery/blob/master/demo.ipynb
In the last part, your case is satisfied.

![image](https://user-images.githubusercontent.com/49093246/232693259-a7dccee9-694b-4880-bee8-1e9885c0d443.png)

For a whole sentence, some obvious texts may take the lead, thus overlook the rest texts (you can draw the similarity map).
So, my suggestion is to treat different text individually, or adjust the threshold to pick points (e.g. 0.8->0.7)

**Comment by xiexie123** [comment]
> 
for question 3.  there is no trainning referred to the papre
